sound/soc/sof/topology.c

```diff
@@ -2194,48 +2202,80 @@ static int sof_link_dmic_load(struct snd_soc_component *scomp, int index,
 	return ret;
 }
 
+/* for hda link, playback and capture are supported by different dai
+ * in FW. Here get the dai_index, set dma channel of each dai
+ * and send config to FW. In FW, each dai sets config by dai_index
+ */
 static int sof_link_hda_process(struct snd_sof_dev *sdev,
 				struct snd_soc_dai_link *link,
 				struct sof_ipc_dai_config *config,
-				int slot,
-				int direction)
+				int tx_slot,
+				int rx_slot)
 {
 	struct sof_ipc_reply reply;
 	u32 size = sizeof(*config);
-	struct snd_sof_dai *dai;
+	struct snd_sof_dai *sof_dai;
 	int found = 0;
 	int ret;
 
-	/* for hda link, playback and capture are supported by different
-	 * dai in FW. Here get the dai_index of each dai and send config
-	 * to FW. In FW, each dai sets config by dai_index
-	 */
-	list_for_each_entry(dai, &sdev->dai_list, list) {
-		if (!dai->name)
+	list_for_each_entry(sof_dai, &sdev->dai_list, list) {
+		if (!sof_dai->name)
 			continue;
 
-		if (strcmp(link->name, dai->name) == 0 &&
-		    dai->comp_dai.direction == direction) {
-			config->dai_index = dai->comp_dai.dai_index;
+		if (strcmp(link->name, sof_dai->name) == 0) {
+			if (sof_dai->comp_dai.direction ==
+			    SNDRV_PCM_STREAM_PLAYBACK) {
+				if (!link->dpcm_playback)
+					return -EINVAL;
+
+				config->hda.link_dma_ch = tx_slot;
+			} else {
+				if (!link->dpcm_capture)
+					return -EINVAL;
+
+				config->hda.link_dma_ch = rx_slot;
+			}
+
+			if (ret < 0) {
```

Comment:
ret is not set before testing it at each iteration, what are you trying to check here?
This looks like a mistake.

Reply:
I am very sorry to have this error. I will remove it. I change my code according to reviewer's comment, but here I missed to delete it.